### PR TITLE
Refactor freqs_cis slice to be safer for PP

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #318
* #322
* __->__ #321

Unchanged: we precompute freqs_cis for max_seqlen, >> seqlen for a given
batch.

Changed: instead of slicing self.freqs_cis down to seqlen at top level
transformer based on the input token shape, we slice it down to seqlen
inside a transformer layer after we have re-expanded to the full seqlen
in cases where TP has sharded across seqlen.

In the PP case, stage 1's input may be seqlen/TP instead of seqlen, but
we do not generally know this.  That makes it hard for stage1 to slice
freqs_cis correctly.  It's easy to do the slicing deeper inside, since
at that point we do know the full seqlen unambiguously.

Note: the full self.freqs_cis is stored in memory either way, and the
thing passed into every layer is just a view. This change should not be
material for memory usage or otherwise.

### CODE REVIEW

**Review by lessw2020 — APPROVED** [review]
makes sense - lgtm!

**Review by wanchaol — APPROVED** [review]
lgtm!

File: torchtitan/models/llama/model.py [hunk 1]
```diff
@@ -76,7 +79,9 @@ def reshape_for_broadcast(freqs_cis: torch.Tensor, x: torch.Tensor) -> torch.Ten
     """
     ndim = x.ndim
     assert 0 <= 1 < ndim
```
Comment (awgu) [inline torchtitan/models/llama/model.py:81]: not from this PR: I wonder what the point of the `0 <= 1` part is 😃 .

Comment (wconstab) [reply torchtitan/models/llama/model.py:81]: lol.  its always good to check your assumptions

**Pull request merged**